File: emulinker/src/main/java/org/emulinker/kaillera/controller/v086/protocol/QuitGame_Notification.java
package org.emulinker.kaillera.controller.v086.protocol;

import com.google.auto.value.AutoValue;
import org.emulinker.kaillera.controller.messaging.MessageFormatException;

@AutoValue
public abstract class QuitGame_Notification extends QuitGame {
  private static final String DESC = "Quit Game Notification";

  public static AutoValue_QuitGame_Notification create(
      int messageNumber, String username, int userId) throws MessageFormatException {
    V086Message.validateMessageNumber(messageNumber, DESC);

    if (userId < 0 || userId > 0xFFFF) {
      throw new MessageFormatException(
          "Invalid " + DESC + " format: userID out of acceptable range: " + userId);
    }
    return new AutoValue_QuitGame_Notification(messageNumber, ID, DESC, username, userId);
  }
}

File: emulinker/src/main/java/org/emulinker/kaillera/controller/v086/protocol/QuitGame_Notification.kt
package org.emulinker.kaillera.controller.v086.protocol

import org.emulinker.kaillera.controller.messaging.MessageFormatException
import org.emulinker.kaillera.controller.v086.protocol.V086Message.Companion.validateMessageNumber

data class QuitGame_Notification
    @Throws(MessageFormatException::class)
    constructor(
        override val messageNumber: Int, override val username: String, override val userId: Int
    ) : QuitGame() {

  init {
    validateMessageNumber(messageNumber)
    require(userId in 0..0xFFFF) { "UserID out of acceptable range: $userId" }
  }
}
